File: preview/preview.go
// Package preview demonstrates all Go syntax elements.
package preview

import (
	"context"
	"fmt"
	"math"
	"strings"
	"sync"
	"time"
)

// Status represents a task state.
type Status int

const (
	Pending  Status = iota // pending
	Running                // running
	Done                   // done
	maxRetry        = 3
	piValue         = 3.14159
	hexFlag         = 0xFF
	greeting        = "hello, world"
)

// Task describes a unit of work.
type Task struct {
	ID        int64             `json:"id"`
	Name      string            `json:"name"`
	Status    Status            `json:"status"`
	Tags      []string          `json:"tags,omitempty"`
	Meta      map[string]any    `json:"meta"`
	CreatedAt time.Time         `json:"created_at"`
	cancel    context.CancelFunc
}

// Runner is an executor interface.
type Runner interface {
	Run(ctx context.Context, task *Task) error
	Stop() error
}

// Scheduler manages tasks.
type Scheduler struct {
	mu      sync.RWMutex
	tasks   []*Task
	workers int
	done    chan struct{}
}

// NewScheduler creates a scheduler.
func NewScheduler(workers int) *Scheduler {
	return &Scheduler{
		workers: workers,
		done:    make(chan struct{}),
	}
}

// Submit adds a task.
func (s *Scheduler) Submit(name string, tags ...string) *Task {
	s.mu.Lock()
	defer s.mu.Unlock()

	t := &Task{
		ID:        int64(len(s.tasks) + 1),
		Name:      name,
		Status:    Pending,
		Tags:      tags,
		Meta:      make(map[string]any),
		CreatedAt: time.Now(),
	}
	s.tasks = append(s.tasks, t)
	return t
}

// Process handles tasks with retries.
func (s *Scheduler) Process(ctx context.Context) error {
	for i, task := range s.tasks {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		task.Status = Running

		for retry := 0; retry < maxRetry; retry++ {
			if err := s.execute(task); err != nil {
				fmt.Printf("task %d/%d error (attempt %d): %v\n",
					i+1, len(s.tasks), retry+1, err)
				continue
			}
			break
		}

		task.Status = Done
	}

	// String operations
	names := make([]string, 0, len(s.tasks))
	for _, t := range s.tasks {
		names = append(names, t.Name)
	}
	summary := strings.Join(names, ", ")
	fmt.Printf("completed: [%s]\n", summary)

	// Numbers and math
	angle := math.Pi / 4
	sin := math.Sin(angle)
	_ = sin

	return nil
}

func (s *Scheduler) execute(t *Task) error {
	if t == nil {
		return fmt.Errorf("task must not be nil")
	}
	t.Meta["executed_at"] = time.Now().Unix()
	return nil
}

// Generics
func Filter[T any](slice []T, predicate func(T) bool) []T {
	result := make([]T, 0)
	for _, item := range slice {
		if predicate(item) {
			result = append(result, item)
		}
	}
	return result
}

// Goroutine + channel
func fanOut(input <-chan int, workers int) <-chan int {
	out := make(chan int)
	var wg sync.WaitGroup
	for w := 0; w < workers; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for n := range input {
				out <- n * n
			}
		}()
	}
	go func() {
		wg.Wait()
		close(out)
	}()
	return out
}

File: preview/preview.rs
// Демонстрация синтаксиса Rust для проверки темы.

use std::collections::HashMap;
use std::fmt;
use std::sync::{Arc, Mutex};

/// Статус задачи.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Status {
    Pending,
    Running,
    Done,
    Failed(u8),
}

/// Задача с метаданными.
#[derive(Debug)]
pub struct Task {
    pub id: u64,
    pub name: String,
    pub status: Status,
    tags: Vec<String>,
    meta: HashMap<String, serde_json::Value>,
}

const MAX_RETRY: u8 = 3;
const PI: f64 = 3.14159_265_358_979;
const HEX_MASK: u32 = 0xFF00_FF00;

impl Task {
    pub fn new(id: u64, name: &str) -> Self {
        Self {
            id,
            name: name.to_string(),
            status: Status::Pending,
            tags: Vec::new(),
            meta: HashMap::new(),
        }
    }

    pub fn with_tags(mut self, tags: &[&str]) -> Self {
        self.tags = tags.iter().map(|s| s.to_string()).collect();
        self
    }

    fn execute(&mut self) -> Result<(), TaskError> {
        if self.name.is_empty() {
            return Err(TaskError::EmptyName);
        }
        self.status = Status::Running;
        self.meta.insert("attempt".into(), serde_json::json!(1));
        self.status = Status::Done;
        Ok(())
    }
}

impl fmt::Display for Task {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Task#{} '{}' [{:?}]", self.id, self.name, self.status)
    }
}

/// Ошибки задачи.
#[derive(Debug, thiserror::Error)]
pub enum TaskError {
    #[error("пустое имя задачи")]
    EmptyName,
    #[error("превышен лимит ретраев ({0})")]
    RetryExhausted(u8),
    #[error("внутренняя ошибка: {0}")]
    Internal(String),
}

/// Планировщик с потокобезопасным доступом.
pub struct Scheduler {
    tasks: Arc<Mutex<Vec<Task>>>,
    workers: usize,
}

impl Scheduler {
    pub fn new(workers: usize) -> Self {
        Self {
            tasks: Arc::new(Mutex::new(Vec::new())),
            workers,
        }
    }

    pub fn submit(&self, name: &str) -> u64 {
        let mut tasks = self.tasks.lock().unwrap();
        let id = tasks.len() as u64 + 1;
        tasks.push(Task::new(id, name));
        id
    }

    pub fn process(&self) -> Result<Vec<String>, TaskError> {
        let mut tasks = self.tasks.lock().unwrap();
        let mut completed = Vec::new();

        for task in tasks.iter_mut() {
            for retry in 0..MAX_RETRY {
                match task.execute() {
                    Ok(()) => {
                        completed.push(format!("{}", task));
                        break;
                    }
                    Err(e) if retry < MAX_RETRY - 1 => {
                        eprintln!("retry {}: {}", retry + 1, e);
                        task.status = Status::Failed(retry);
                        continue;
                    }
                    Err(e) => return Err(e),
                }
            }
        }

        Ok(completed)
    }
}

// Generics + trait bounds
fn filter<T, F>(items: &[T], predicate: F) -> Vec<&T>
where
    F: Fn(&T) -> bool,
{
    items.iter().filter(|item| predicate(item)).collect()
}

// Closures, iterators, pattern matching
fn summary(tasks: &[Task]) -> String {
    tasks
        .iter()
        .filter_map(|t| match t.status {
            Status::Done => Some(t.name.as_str()),
            Status::Failed(n) => Some(if n > 1 { "failed" } else { "retry" }),
            _ => None,
        })
        .collect::<Vec<_>>()
        .join(", ")
}

// Async
async fn fetch_task(id: u64) -> Result<Task, TaskError> {
    let task = Task::new(id, &format!("async-task-{}", id));
    tokio::time::sleep(std::time::Duration::from_millis(100)).await;
    Ok(task)
}
